system/lib/pthread/emscripten_tls_init.c

```diff
@@ -28,7 +28,12 @@ static void free_tls(void* tls_block) {
   emscripten_builtin_free(tls_block);
 }
 
-// Note that ASan constructor priority is 50, and we must be higher.
+// The current sequence of system-level static init functions in emscripten is:
+// 47: emmalloc_init
+// 48: __emscripten_pthread_data_constructor
+// 49: emscripten_tls_init
+// 50: asan init
+// (Please keep the above list updated alongside each one of these fuctions).
```

Comment:
Really good to have a documentation like this. I wonder if we should put this snippet in a separate text file, and have this comment instead say something like

```
The global sequence of system-level static initialization ctor orders is coordinated in the file docs/system_ctor_order.txt. If changing the constructor ordering, please update that file as well.
```

Does the order number 50 have some special meaning for asan e.g. as far as other native platforms are considered? Or are we free to adjust the order as we please?

It would be good to have the file docs/system_ctor_order.txt document the usable range of ctor numbers allocated for end users? E.g.
```
//  1-40: safe ctor numbers for Emscripten users to utilize.
```

Or maybe we could separate the numbers we use always by two at least. E.g. only use odd numbers by Emscripten, and reserve even numbers for end users. That way we will have a guarantee that end users can always order between, say, emmalloc and tls.

(That may all be over-engineering for a non-problem scenario, but in practice a simple convention to take, so maybe we might as well do something like that?)

Reply:
Apparently numbers 0-100 are reserved in theory for system constructors:

https://www.keil.com/support/man/docs/armclang_ref/armclang_ref_chr1383738546439.htm

This means constructors (such as C++ static constructors) should always be > 100.     However this is not really enforceable which is why I'd like to move away from it and do something more static (with weak symbol references perhaps) eventually. 

One example of use the of user priorities is C++ IO streams which get initialized at 101 which is the highest priority user-level constructor:

https://github.com/emscripten-core/emscripten/blob/d35de0bf3572dfe9759ac3c7e56b9ce5983f68ac/system/lib/libcxx/include/__config#L1448